### PR TITLE
[Access] Add flag to reindex execution data

### PR DESCRIPTION
**Pull request by peterargue**
Closes: https://github.com/onflow/flow-go/issues/4939

This PR adds a new flag to Access nodes (`--execution-data-force-reset-indexer-height`) that allows an operator to reindex execution data starting from the root block. This is helpful when a bug is discovered, or the data is found to be corrupt.

Steps to recover:
1. Stop the node
2. Backup the `execution-state` and `protocol` data directories
3. Delete the `execution-state`. DO NOT delete the protocol db.
4. Set `--execution-data-force-reset-indexer-height=true` in node's startup config
5. Start the node
6. Remove `--execution-data-force-reset-indexer-height` from the startup config so it's not used on future reboots. You do not need to restart the node again.

The node will load the checkpoint into the execution-state db, then reset the indexer's current height value, causing it to reindex from the root block.

Note: This is OK with the current node software because the events/tx results data indexed from execution data is not used. Eventually, we will start to use it in the API, at which point we may need another method to ensure it is not returned in queries while reindexing.

### CODE REVIEW

File: cmd/access/node_builder/access_node_builder.go [hunk 1]
```diff
@@ -722,6 +723,15 @@ func (builder *FlowAccessNodeBuilder) BuildExecutionSyncComponents() *FlowAccess
 					if err != nil {
 						return nil, fmt.Errorf("could not load checkpoint file: %w", err)
 					}
+
+					// this is only supported when bootstrapping a new register db, otherwise the
+					// register db will throw exceptions when indexing an already indexed height.
+					if builder.forceResetIndexerHeight {
```
Comment (sideninja) [inline cmd/access/node_builder/access_node_builder.go:729]: shouldn't this actually always be set like that? is there a case where we would only remove the checkpoint content?

**Review by sideninja — COMMENTED** [review]
left a question

File: cmd/access/node_builder/access_node_builder.go (hunk 1 above)
Comment (peterargue) [reply cmd/access/node_builder/access_node_builder.go:729]: currently, we can only remove the register data. events, results, etc, are stored in the protocol db, and can't be removed without some more code.

This requirement is here because there's a race condition between setting the register `latest` height and the indexer's current height, where it's possible the register height is updated first. In that case, data could be missed if someone rebooted with this flag enabled and an existing db.

File: cmd/access/node_builder/access_node_builder.go [hunk 2]
```diff
@@ -722,6 +723,15 @@ func (builder *FlowAccessNodeBuilder) BuildExecutionSyncComponents() *FlowAccess
...
+						err = indexedBlockHeight.SetProcessedIndex(checkpointHeight)
+						if err != nil && !errors.Is(err, storage.ErrNotFound) {
+							return nil, fmt.Errorf("could not reset indexed block height: %w", err)
+						}
```
Comment (zhangchiqing) [inline cmd/access/node_builder/access_node_builder.go:733]: we should print a log before doing the operation and after it's done.

File: cmd/access/node_builder/access_node_builder.go [hunk 3]
```diff
@@ -722,6 +723,15 @@ func (builder *FlowAccessNodeBuilder) BuildExecutionSyncComponents() *FlowAccess
 					if err != nil {
 						return nil, fmt.Errorf("could not load checkpoint file: %w", err)
 					}
+
+					// this is only supported when bootstrapping a new register db, otherwise the
+					// register db will throw exceptions when indexing an already indexed height.
+					if builder.forceResetIndexerHeight {
+						err = indexedBlockHeight.SetProcessedIndex(checkpointHeight)
```
Comment (zhangchiqing) [inline cmd/access/node_builder/access_node_builder.go:730]: before resetting the index, it's better check the pebble database is empty, otherwise, the pebble DB will be built from a dirty state, and I'm not sure if the register index will behave correctly.

File: cmd/access/node_builder/access_node_builder.go (hunk 2 above)
Comment (peterargue) [reply cmd/access/node_builder/access_node_builder.go:734]: addressed in [0d10562](https://github.com/onflow/flow-go/pull/4943/commits/0d1056235c21362f3f02caa8cefcb0150a109b91)

File: cmd/access/node_builder/access_node_builder.go (hunk 3 above)
Comment (zhangchiqing) [reply cmd/access/node_builder/access_node_builder.go:731]: OK, maybe we should add a check that if pebble DB is not bootstrapped, then the indexedBlockHeight should have no ProcessedIndex stored, otherwise it means the database is corrupted. 

So the database is corrupted if:
1. pebble DB is bootstrapped, but processed index is not set
2. pebble DB is not bootstrapped, but processed index is set

In case 1, we should probably add a check.
In case 2, if forceResetIndexerHeight == checkpointHeight, then we will reset the indexer height, otherwise, we should crash.

File: cmd/access/node_builder/access_node_builder.go [hunk 4]
```diff
@@ -722,6 +723,17 @@ func (builder *FlowAccessNodeBuilder) BuildExecutionSyncComponents() *FlowAccess
 					if err != nil {
 						return nil, fmt.Errorf("could not load checkpoint file: %w", err)
 					}
+
+					// this is only supported when bootstrapping a new register db, otherwise the
+					// register db will throw exceptions when indexing an already indexed height.
+					if builder.forceResetIndexerHeight {
```
Comment (zhangchiqing) [inline cmd/access/node_builder/access_node_builder.go:729]: Maybe better to use a height, so that the operator doesn't need to change it back to false after the reindex.
```suggestion
					shouldResetIndexHeight := builder.forceResetIndexerHeight == checkpointHeight 
					if shouldResetIndexHeight {
```

Comment (peterargue) [reply cmd/access/node_builder/access_node_builder.go:729]: this would require them to set the right height though. As is, this logic is only executed if the registers db is not bootstrapped, so it should never be called again on subsequent reboots.

File: cmd/access/node_builder/access_node_builder.go (hunk 3 above)
Comment (peterargue) [reply cmd/access/node_builder/access_node_builder.go:731]: 1. could happen if the node crashed at some point during startup. we would have to check if pebble is bootstrapped and the register db's latest height is != root height and processed index is unset.

2. makes sense

Comment (peterargue) [reply cmd/access/node_builder/access_node_builder.go:769]: @zhangchiqing I added some checks and comments in [037a811](https://github.com/onflow/flow-go/pull/4943/commits/037a81195f70aaae364a694c0da0261ff6a54145). Still need to look into testing. Does this address your concern?

**Review by sideninja — APPROVED** [review]

File: cmd/access/node_builder/access_node_builder.go [hunk 5]
```diff
@@ -690,7 +691,45 @@ func (builder *FlowAccessNodeBuilder) BuildExecutionSyncComponents() *FlowAccess
...
+					// execution-state db was kept. neither case is valid
+					registers, err = pStorage.NewRegisters(pdb)
+					if err != nil {
+						return nil, fmt.Errorf("could not create registers storage: %w", err)
+					}
+
+					// the registers db may have been bootstrapped, but the node crashed before
+					// initializing the indexer. In this case, the register db's latest height
+					// will be the root height.
+					if !indexerInitialized && registers.LatestHeight() != builder.SealedRootBlock.Header.Height {
```
Comment (zhangchiqing) [inline cmd/access/node_builder/access_node_builder.go:718]: I think you are adding the check that we discussed. However, this logic is getting quite complex. And I don't quite understand why we need to check the result of `&& registers.LatestHeight() != builder.SealedRootBlock.Header.Height` .

I think it's better to be moved to a separate function/module. This piece of logic is checking whether the indexer's dependencies are in a consistent state, should not be the responsibility of the node builder. Also when this check is in the access_node_builder, it's hard to write test cases for them.

IMO, the best place for this check is in the indexer itself, because it's the indexer itself updates the states of the dependencies, it has the best knowledge of what a consistent state should be. And in our test case, we could check the dependencies must be in a consistent state after certain operations, such as finish indexing a new block. 

In other words, the constructor could take all the dependencies,  flags andconfigs, then check if dependencies are consistency, and make adjustment (force reset index), after everything is done, then return the constructed instance.

**Comment by peterargue** [comment]
refactored the logic into the indexer module. still need to add tests

File: cmd/access/node_builder/access_node_builder.go [hunk 6]
```diff
@@ -938,6 +940,7 @@ func (builder *FlowAccessNodeBuilder) extraFlags() {
 		flags.BoolVar(&builder.executionDataIndexingEnabled, "execution-data-indexing-enabled", defaultConfig.executionDataIndexingEnabled, "whether to enable the execution data indexing")
 		flags.StringVar(&builder.registersDBPath, "execution-state-dir", defaultConfig.registersDBPath, "directory to use for execution-state database")
 		flags.StringVar(&builder.checkpointFile, "execution-state-checkpoint", defaultConfig.checkpointFile, "execution-state checkpoint file")
+		flags.BoolVar(&builder.forceResetIndexerHeight, "execution-data-force-reset-indexer-height", false, "reset indexer processed height to root block. DO NOT use unless you know what you are doing")
```
Comment (sideninja) [inline cmd/access/node_builder/access_node_builder.go:943]: 53ea296e-9389-11ed-ad11-0210609a3fe2
